File: churn-demo/new/ledger.cpp
#include <cmath>
#include <string>
#include <vector>
#include <map>

struct Entry {
    std::string status;
    double amount;
};

class Ledger {
public:
    Ledger(const std::string& name) : name_(name), balance_(0.0), limit_(0.0) {}

    void post(const std::string& status, double amount) {
        entries_.push_back(Entry{status, amount});
        balance_ = balance_ + amount;
        log("posted");
    }

    void credit(double amount) {
        post("posted", amount);
        clearedItems_ += 1;
    }

    void debit(double amount) {
        post("posted", -amount);
        clearedItems_ += 1;
    }

    void hold(double amount) {
        post("pending", amount);
        pendingHolds_ += 1;
    }

    void dispute(double amount) {
        post("disputed", amount);
        disputedItems_ += 1;
        flaggedItems_ += 1;
    }

    void transfer(Ledger& to, double amount) {
        debit(amount);
        to.credit(amount);
        transfers_ += 1;
    }

    void applyInterest(double rate) {
        double interest = balance_ * rate;
        post("posted", interest);
        interestPaid_ += interest;
    }

    int flagLarge(double threshold) {
        int n = 0;
        for (const auto& e : entries_) {
            if (std::abs(e.amount) > threshold) n += 1;
        }
        flaggedItems_ += n;
        return n;
    }

    double largestEntry() const {
        double best = 0.0;
        for (const auto& e : entries_) {
            if (std::abs(e.amount) > std::abs(best)) best = e.amount;
        }
        return best;
    }

    std::string statementLine(int i) const {
        const Entry& e = entries_.at(i);
        return e.status + ":" + std::to_string(e.amount);
    }

    void rename(const std::string& newName) {
        previousName_ = name_;
        name_ = newName;
        log("renamed");
    }

    double netFlow() const {
        return sumSigned(+1) - sumSigned(-1);
    }

    void reset() {
        entries_.clear();
        summary_.clear();
        balance_ = 0.0;
        log("reset");
    }

    double sumSigned(int sign) const {
        double sum = 0.0;
        for (const auto& e : entries_) {
            if ((e.amount >= 0) == (sign > 0)) sum += std::abs(e.amount);
        }
        return sum;
    }

    int countStatus(const std::string& status) const {
        int n = 0;
        for (const auto& e : entries_) {
            if (e.status == status) n += 1;
        }
        return n;
    }

    void setLimit(double limit) {
        limit_ = limit;
        log("limit changed");
    }

    bool overLimit() const {
        return balance_ < -limit_;
    }

    std::string name() const {
        return name_;
    }

    double balance() const {
        return balance_;
    }

    int entryCount() const {
        return static_cast<int>(entries_.size());
    }

    double summaryValue(const std::string& key) const {
        auto it = summary_.find(key);
        return it == summary_.end() ? 0.0 : it->second;
    }

    void reconcile() {
        double opening = balance_;
        double credits = sumSigned(+1);
        double debits = sumSigned(-1);
        double closing = opening + credits - debits;
        double drift = closing - balance_;
        int posted = countStatus("posted");
        int pending = countStatus("pending");
        int disputed = countStatus("disputed");
        int total = posted + pending + disputed;
        double average = total > 0 ? (credits + debits) / total : 0.0;
        summary_["opening"] = opening;
        summary_["credits"] = credits;
        summary_["debits"] = debits;
        summary_["closing"] = closing;
        summary_["drift"] = drift;
        summary_["posted"] = posted;
        summary_["pending"] = pending;
        summary_["disputed"] = disputed;
        summary_["average"] = average;
        balance_ = closing;
    }

    void resetCounters() { pendingHolds_ = 0; clearedItems_ = 0; flaggedItems_ = 0; disputedItems_ = 0; }

private:
    void log(const std::string& m) const {}
    std::string name_;
    std::string previousName_;
    double balance_;
    double limit_;
    double interestPaid_ = 0.0;
    int transfers_ = 0;
    int pendingHolds_ = 0;
    int clearedItems_ = 0;
    int flaggedItems_ = 0;
    int disputedItems_ = 0;
    std::vector<Entry> entries_;
    std::map<std::string, double> summary_;
};

File: churn-demo/old/ledger.js
class Ledger {
    constructor(name) {
        this.name = name;
        this.previousName = "";
        this.balance = 0;
        this.limit = 0;
        this.interestPaid = 0;
        this.transfers = 0;
        this.pendingHolds = 0;
        this.clearedItems = 0;
        this.flaggedItems = 0;
        this.disputedItems = 0;
        this.entries = [];
        this.summary = new Map();
        this.lastLog = "";
    }

    log(m) {
        this.lastLog = m;
    }

    reconcile() {
        const opening = this.balance;
        const credits = this.sumSigned(+1);
        const debits = this.sumSigned(-1);
        const closing = opening + credits - debits;
        const drift = closing - this.balance;
        const posted = this.countStatus("posted");
        const pending = this.countStatus("pending");
        const disputed = this.countStatus("disputed");
        const total = posted + pending + disputed;
        const average = total > 0 ? (credits + debits) / total : 0;
        this.summary.set("opening", opening);
        this.summary.set("credits", credits);
        this.summary.set("debits", debits);
        this.summary.set("closing", closing);
        this.summary.set("drift", drift);
        this.summary.set("posted", posted);
        this.summary.set("pending", pending);
        this.summary.set("disputed", disputed);
        this.summary.set("average", average);
        this.balance = closing;
    }

    resetCounters() { this.pendingHolds = 0; this.clearedItems = 0; this.flaggedItems = 0; this.disputedItems = 0; }

    post(status, amount) {
        this.entries.push({ status, amount });
        this.balance += amount;
        this.log("posted");
    }

    credit(amount) {
        this.post("posted", amount);
        this.clearedItems += 1;
    }

    debit(amount) {
        this.post("posted", -amount);
        this.clearedItems += 1;
    }

    hold(amount) {
        this.post("pending", amount);
        this.pendingHolds += 1;
    }

    dispute(amount) {
        this.post("disputed", amount);
        this.disputedItems += 1;
        this.flaggedItems += 1;
    }

    transfer(to, amount) {
        this.debit(amount);
        to.credit(amount);
        this.transfers += 1;
    }

    applyInterest(rate) {
        const interest = this.balance * rate;
        this.post("posted", interest);
        this.interestPaid += interest;
    }

    flagLarge(threshold) {
        let n = 0;
        for (const e of this.entries) {
            if (Math.abs(e.amount) > threshold) n += 1;
        }
        this.flaggedItems += n;
        return n;
    }

    largestEntry() {
        let best = 0;
        for (const e of this.entries) {
            if (Math.abs(e.amount) > Math.abs(best)) best = e.amount;
        }
        return best;
    }

    statementLine(i) {
        const e = this.entries[i];
        return e.status + ":" + String(e.amount);
    }

    rename(newName) {
        this.previousName = this.name;
        this.name = newName;
        this.log("renamed");
    }

    netFlow() {
        return this.sumSigned(+1) - this.sumSigned(-1);
    }

    reset() {
        this.entries = [];
        this.summary = new Map();
        this.balance = 0;
        this.log("reset");
    }

    sumSigned(sign) {
        let sum = 0;
        for (const e of this.entries) {
            if ((e.amount >= 0) === (sign > 0)) sum += Math.abs(e.amount);
        }
        return sum;
    }

    countStatus(status) {
        let n = 0;
        for (const e of this.entries) {
            if (e.status === status) n += 1;
        }
        return n;
    }

    setLimit(limit) {
        this.limit = limit;
        this.log("limit changed");
    }

    overLimit() {
        return this.balance < -this.limit;
    }

    entryCount() {
        return this.entries.length;
    }

    summaryValue(key) {
        return this.summary.get(key) ?? 0;
    }
}

module.exports = { Ledger };
